### ISSUE TITLE
Consider making `fsharp_experimental_stroustrup_style` apply independently of `fsharp_multiline_block_brackets_on_same_column`

### ISSUE THREAD

**Issue by josh-degraw** [fsprojects/fantomas#2425]
As mentioned in a comment in #2413:

> Using `fsharp_experimental_stroustrup_style` without `fsharp_multiline_block_brackets_on_same_column` does not make any sense. ExperimentalStroustrupStyle is an extension of MultilineBlockBracketsOnSameColumn.

Currently, to use the experimental Stroustrup setting you need to enable two flags. Other than the potential issue of not knowing you need to enable two flags, the `MultilineBlockBracketsOnSameColumn` is technically inaccurate when using the Stroustrup style, as by definition the brackets are on different columns.

I propose that:
- If `fsharp_experimental_stroustrup_style` is enabled, `fsharp_multiline_block_brackets_on_same_column` should be treated as redundant and relevant logic should be applied automatically whether it's present or not.
- Logic that would format differently when using Stroustrup vs `MultilineBlockBracketsOnSameColumn` should only need to check if the Stroustrup setting is enabled, 
- Logic that would apply identically for either Stroustrup or `MultilineBlockBracketsOnSameColumn`, but not otherwise, should check if either flag is present instead of only being applied if `MultilineBlockBracketsOnSameColumn` is true.

Alternatively, it might make sense to have a different flag to encapsulate the possible bracket styles (e.g. `stroustrup`, `allman`, `classic`, or whatever is most accurate). Admittedly a change like that could balloon out in logic handling, but for configuring the code it could make it much more straightforward to only need one flag to represent all potential bracket indentation styles.

Regardless, IMHO the biggest issue with the current api is mainly the fact that `MultilineBlockBracketsOnSameColumn` is factually incorrect when Stroustrup style is applied, so I think an adjustment is at least worth discussing due to the potential confusion it could cause.

**Comment by nojaf** [comment]
Hello, thank you for bringing this up. 

> the MultilineBlockBracketsOnSameColumn is technically inaccurate when using the Stroustrup style, as by definition the brackets are on different columns.

That is not entirely true, there are situations when `stroupstrup` isn't applicable and `MultilineBlockBracketsOnSameColumn` is used.
[Sample](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAYrLGALwA6ATmQHYQTCU13MFFoDO6FAltQOaNaLCACEAhhTS90g4RACC1WOgAWiSWE49%2BsugF9dEALaIjAI3VgAHgDoAEogA2j4iTAAKAJSDBjxOjAxUkMwAG0wYFZXMDJIQlgARliAbjBxClIwACYAZlTFZTUMt1jVXgFIMAMqIWZI%2BMzY%2BKyUtIlM3PylVUsSyDKdSuqmFnqiRriiHNb0jrywAp7imP6VctiqkPooifiAFhn2t06F7qKJgYqqsABdHxrIABpwKHZ-dHL2VDCX%2Bl%2B6SAAEiwuG%2BkGkT3%2BECBADcxI4AK7INChKF0AAMzzkUAAktQkNR0ABlbgAL2QWLkezRN3%2Bekpf2EQJByKgEIZAMBcMRrNR2ISHOYkAAsmIrAAZXiIcWIfiqSH8rLoml0jmgJnAnCsyCy%2BAAeQAZpLqIgiVg-AqWLD4Ujvny5FlBQCAKL4w3Gilo6GOA2QFXMen-dVWzWgtCQMxERyWoVcm28r1gHJO6F4t4UdD4XjwgByiAA7o4pTGWFwkf79Gq0cytWDI7Bo07rTy7Ym9imoETsGJoIgRIgDbAKIgAAoSMQmdDqEvMMuICsQQN1auh7X1xsr7m2lGJgCsHcgXZ7fYHQ%2BlsHz6mgYjeeJhsGvH1g1BndDnC6qVY1LLrUdfze3H5sQANgPI9e37QdhwAVWwXAKGvW9qHvR9uGfV8IANeE3g-Jc6GDWMf3Ddd-zjFsd2xAB2MDuwg09hwAYUcG92AY58tARaB0CHDCwCwxwcOEWkAy-EMiKgEimzIwD7WEAAOGjj0gs9hVMCwKF4-jBJYYTKyDFdxIjP8pK3BNsQATkUuioMQNiXBfDstPnITVX079a2I4zN3jVt%2BUxRND1oxB5ANKcKDYowjDEXj3xckS3LEjyJK8jVTN8uQEgFALwJPGyiVMbg4HszTsOcnTXOXdyw2ShtSLSiiMsdbKgpCsL8qMQqG3QjtYvK%2BLKsS6qjNqkyfIa4QEmTZrj3kChYARfEABEnG4DqwpiihyzivSBsIpLwUJOqxqAjL2wC0UrBxA0ABU1GoIkVCHdAAHVuHgeUO2VbbF1EvahvZbzyJOib93OsUrtu2VnQE01HozV73pUXjgK%2Bvqdvwgz9oB1LjtklgElAsHLuoA1uCsPV4LEbiKGdKxsGHdh2DQhzEzk1HmF0n6Er%2B7VsZDergfx6iiYAJUQOAKHgBGPrbdm6E5z9uc5Qy%2BdjAW8eYBIFNF8Wh3gHMEXMdRDRxKcjC%2BDsElw37lf2owEUcD4ixNQgKCi9Awuu2tRqBjW6ASSyArFiX4GFB2nalV33fWg9oBUccuPUAB9fM3vla2lehVcwVVzl1cTJUDwu2aKDETA9QoSVOGlpGOzZjPdtt-7Dp9mSC6y7ERTFEuy4rqv0ANo2KBNs2LcTK3vsVxus8M%2B3He4Z3ECjqnPe9wG2%2BxLIms7nvy8r7hODD%2BfF%2BXj3p1j%2BPS8TigU7TpGG4xqreZb9ezIdKbO4ugA1eMRF4eByg114vXSeeFGSDWfugI6vsC5nU-mKfAC0uLMz-viQBd9eJ7DlhABWYCIiY2blA1ub9hBZFBvAqwi1lAAHF-C03pogRmzMgF12wbcCqj8IG-hGq-dKpDCaf3DgvKUIgXDQAANYiCvuI-w7A9T3QnLZBshsWbYl6hzDh4CebcI3DjGBm9hadzzIWER-hLyyi9rgZapNqDcCfNQRQoc1LqDHmozaZUNH9U4dozyPC9EbwdNrHeRY%2BDUEQdQZBz4SShKpgiYc11YB4gJOgKmzMSowwflopua4Ur81xgXQOwSwrUFSTCc8-BVJD2sbwOxzNXFyCcpk-BT8dHQICcIHI-lBHzw9OKCcZh4BiCYrAJm-BjGL3SdpTx6Mskz32pJXh40WA5A7nISA9D1CrVlCkxwABpRAiBsBJO2XiKRYgIm10TI00BNs5lDQWf4khyzt5rNEec8RHp2CzXmviPMnBEChyEYvehDMmbsQ2ltNGXNp5QGzr43ReT9FyByB-V5EhlLDkWgfaAPAOolKnPAaCtjnzLWgMxUu9jJkePlpo5pXD4VtKecwHIcC1kbNxds%2BEZo5oIg4tgM0mALQ9XcU0gi2Sc4v0eXw5Z5C1n7MORdQe6lDRvOoB8qU9SJrojYbg25sLDIPMRe05ZAi1ncsKugYUCBPTYmuVCqoLwbiUnBOwfATNviNL0EAA)

Stroustrup isn't [documented](https://fsprojects.github.io/fantomas/docs/end-users/Configuration.html) I see. I would accept a PR for that but as we are already in beta I don't have any desire to change the public API.

Lastly, as the setting is forbidden fruit I can very much live with the very minor UX inconvennience.
Every setting is currently independent from the others. I have no interest in shady constructions if setting A then setting B no longer counts and whatnot.

**Comment by josh-degraw** [comment]
> That is not entirely true, there are situations when `stroupstrup` isn't applicable and `MultilineBlockBracketsOnSameColumn` is used.
> [Sample](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAYrLGALwA6ATmQHYQTCU13MFFoDO6FAltQOaNaLCACEAhhTS90g4RACC1WOgAWiSWE49%2BsugF9dEALaIjAI3VgAHgDoAEogA2j4iTAAKAJSDBjxOjAxUkMwAG0wYFZXMDJIQlgARliAbjBxClIwACYAZlTFZTUMt1jVXgFIMAMqIWZI%2BMzY%2BKyUtIlM3PylVUsSyDKdSuqmFnqiRriiHNb0jrywAp7imP6VctiqkPooifiAFhn2t06F7qKJgYqqsABdHxrIABpwKHZ-dHL2VDCX%2Bl%2B6SAAEiwuG%2BkGkT3%2BECBADcxI4AK7INChKF0AAMzzkUAAktQkNR0ABlbgAL2QWLkezRN3%2Bekpf2EQJByKgEIZAMBcMRrNR2ISHOYkAAsmIrAAZXiIcWIfiqSH8rLoml0jmgJnAnCsyCy%2BAAeQAZpLqIgiVg-AqWLD4Ujvny5FlBQCAKL4w3Gilo6GOA2QFXMen-dVWzWgtCQMxERyWoVcm28r1gHJO6F4t4UdD4XjwgByiAA7o4pTGWFwkf79Gq0cytWDI7Bo07rTy7Ym9imoETsGJoIgRIgDbAKIgAAoSMQmdDqEvMMuICsQQN1auh7X1xsr7m2lGJgCsHcgXZ7fYHQ%2BlsHz6mgYjeeJhsGvH1g1BndDnC6qVY1LLrUdfze3H5sQANgPI9e37QdhwAVWwXAKGvW9qHvR9uGfV8IANeE3g-Jc6GDWMf3Ddd-zjFsd2xAB2MDuwg09hwAYUcG92AY58tARaB0CHDCwCwxwcOEWkAy-EMiKgEimzIwD7WEAAOGjj0gs9hVMCwKF4-jBJYYTKyDFdxIjP8pK3BNsQATkUuioMQNiXBfDstPnITVX079a2I4zN3jVt%2BUxRND1oxB5ANKcKDYowjDEXj3xckS3LEjyJK8jVTN8uQEgFALwJPGyiVMbg4HszTsOcnTXOXdyw2ShtSLSiiMsdbKgpCsL8qMQqG3QjtYvK%2BLKsS6qjNqkyfIa4QEmTZrj3kChYARfEABEnG4DqwpiihyzivSBsIpLwUJOqxqAjL2wC0UrBxA0ABU1GoIkVCHdAAHVuHgeUO2VbbF1EvahvZbzyJOib93OsUrtu2VnQE01HozV73pUXjgK%2Bvqdvwgz9oB1LjtklgElAsHLuoA1uCsPV4LEbiKGdKxsGHdh2DQhzEzk1HmF0n6Er%2B7VsZDergfx6iiYAJUQOAKHgBGPrbdm6E5z9uc5Qy%2BdjAW8eYBIFNF8Wh3gHMEXMdRDRxKcjC%2BDsElw37lf2owEUcD4ixNQgKCi9Awuu2tRqBjW6ASSyArFiX4GFB2nalV33fWg9oBUccuPUAB9fM3vla2lehVcwVVzl1cTJUDwu2aKDETA9QoSVOGlpGOzZjPdtt-7Dp9mSC6y7ERTFEuy4rqv0ANo2KBNs2LcTK3vsVxus8M%2B3He4Z3ECjqnPe9wG2%2BxLIms7nvy8r7hODD%2BfF%2BXj3p1j%2BPS8TigU7TpGG4xqreZb9ezIdKbO4ugA1eMRF4eByg114vXSeeFGSDWfugI6vsC5nU-mKfAC0uLMz-viQBd9eJ7DlhABWYCIiY2blA1ub9hBZFBvAqwi1lAAHF-C03pogRmzMgF12wbcCqj8IG-hGq-dKpDCaf3DgvKUIgXDQAANYiCvuI-w7A9T3QnLZBshsWbYl6hzDh4CebcI3DjGBm9hadzzIWER-hLyyi9rgZapNqDcCfNQRQoc1LqDHmozaZUNH9U4dozyPC9EbwdNrHeRY%2BDUEQdQZBz4SShKpgiYc11YB4gJOgKmzMSowwflopua4Ur81xgXQOwSwrUFSTCc8-BVJD2sbwOxzNXFyCcpk-BT8dHQICcIHI-lBHzw9OKCcZh4BiCYrAJm-BjGL3SdpTx6Mskz32pJXh40WA5A7nISA9D1CrVlCkxwABpRAiBsBJO2XiKRYgIm10TI00BNs5lDQWf4khyzt5rNEec8RHp2CzXmviPMnBEChyEYvehDMmbsQ2ltNGXNp5QGzr43ReT9FyByB-V5EhlLDkWgfaAPAOolKnPAaCtjnzLWgMxUu9jJkePlpo5pXD4VtKecwHIcC1kbNxds%2BEZo5oIg4tgM0mALQ9XcU0gi2Sc4v0eXw5Z5C1n7MORdQe6lDRvOoB8qU9SJrojYbg25sLDIPMRe05ZAi1ncsKugYUCBPTYmuVCqoLwbiUnBOwfATNviNL0EAA)

That's a fair counter example, although it's worth noting that that's an exception to the style rule. My main point with that was that it can be confusing to have to specify putting brackets on the same column when the _main purpose_ of Stroustrup style is to have them on different columns, so having the effects of SameColumn apply by default would result in less confusion imho without changing any of the public API.

> Every setting is currently independent from the others. I have no interest in shady constructions if setting A then setting B no longer counts and whatnot.

That's actually kind of my main point, that they're _not_ currently independent. If both settings have to be checked for one to apply, one is very dependent on the other. What I'm suggesting is to make it so that turning on the Stroustrup setting works independently, whether or not you've enabled any other settings. That's all I meant be supercede. A better word choice on my part probably would have been to say that _the effects_ of the same column rule (those that would still be applied such as your example above) should be _implied_ and enabled if Stroustrup is checked, rather than only if both settings are enabled manually.

**Comment by josh-degraw** [comment]
The more I think about it, the more I think it makes the most sense to use a single flag for bracket formatting, since although Stroustrup currently relies a lot on the Allman implementations under the hood, from an end user perspective, "bracket style" can be seen as a mutually exclusive decision between a specific set of options, which I think is why a bunch of folks have gotten confused by the fact that Stroustrup currently requires 2 settings.

In this case, it could be represented by something like:

```fsharp
type BracketStyle =
    | Classic
    | Allman
    | Stroustrup
```

My gut tells me that this might also make the implementation points a bit clearer and potentially simpler because of exhaustive type checking.

I recognize though that this would require a _lot_ of plumbing changes, but I'm willing to toy around with it and see if it's feasible, if it's a direction you're willing to entertain.

**Comment by nojaf** [comment]
I'm willing to go there eventually but not in the short term.
There is still much ground to cover here, I think the entire Stroustrup story (style guide, implementation and documentation) should be very mature before we can move to a bracket-style setting.

Sidenote: We also deliberately didn't call `Allman` `Allman` style in the setting name.
`fsharp_multiline_block_brackets_on_same_column` was a vague common denominator to cover everything listed in the G-Research style guide.

**Comment by josh-degraw** [comment]
> Sidenote: We also deliberately didn't call `Allman` `Allman` style in the setting name.
> `fsharp_multiline_block_brackets_on_same_column` was a vague common denominator to cover everything listed in the G-Research style guide.

Fair enough, I actually agree since it's not "true" Allman (it could also be argued that it's not "true" Stroustrup either haha). But yeah I messed around with it a bit yesterday and I think for a real implementation I would do something like:

```fsharp
type MultilineBracketStyle = 
    | Classic 
    | Aligned 
    | ExperimentalStroustrup
```

For what it's worth, I had some time yesterday and I was able to get an implementation fully working, and I do think at least from my perspective it made it a lot easier to understand the implementation differences between the 3 bracket styles. 

The only major task I still would need to do would be to add a bit of logic to keep the editorconfig parsing backwards compatible, but the actual implementation required fewer changes than I expected to get all the formatting tests to pass. 

I think it can be done in a way that retains full backwards compatibility to avoid needing a breaking change, but still adds some tangible benefits to both the implementation and the end user sides especially in the form of clarity.

**Comment by nojaf** [comment]
Fix available in [v5.2.0-beta-001](https://www.nuget.org/packages/fantomas/5.2.0-beta-001)
